### PR TITLE
Add support for DATE type and TypeCast Transform

### PR DESCRIPTION
**Pull request by haojin2**
This PR adds support for DATE type and transform for TypeCast parsenodes. (Related issues: #909 #852) 
- Added the DATE type to corresponding places.
- Added the transform for TypeCast in Postgres: For type casts, the final cast results will be added to the parse tree (e.g. ' "2017-12-31"::DATE ' will be a single DATE type constant value expression after the transform)
- Added a cast from VARCHAR to DATE type with the default format for date strings: "YYYY-MM-DD".
- Provided an additional fix for handling memory leak in InsertTransform when faced with exceptions.
- Added additional helper functions to transform type names to DataType classes and type::typeId classes.

### CODE REVIEW

File: src/include/parser/create_statement.h [hunk 1]
```diff
@@ -48,23 +48,94 @@ struct ColumnDefinition {
     TEXT,
 
     VARCHAR,
-    VARBINARY
+    VARBINARY,
+
```
Comment (apavlo) [inline src/include/parser/create_statement.h]: Unnecessary newline

File: src/include/parser/create_statement.h [hunk 2]
```diff
@@ -48,23 +48,94 @@ struct ColumnDefinition {
...
+    } else if (strcmp(str, "date") == 0) {
+      data_type = ColumnDefinition::DataType::DATE;
+    } else {
+      throw NotImplementedException(
+          StringUtil::Format("DataType %s not supported yet...\n", str));
+    }
+    return data_type;
+  }
+
+  static type::TypeId StrToValueType(char *str) {
```
Comment (apavlo) [inline src/include/parser/create_statement.h]: How are `StrToDataType` and `StrToValueType` different? They look really similar to me?

Comment (haojin2) [reply src/include/parser/create_statement.h]: StrToDataType is for transforming to DataType, StrToValueType is for transforming typeId.

File: src/include/parser/create_statement.h (hunk 1 above)
Comment (haojin2) [reply src/include/parser/create_statement.h]: Done

**Comment by apavlo** [comment]
I am going to rebase + merge this. I will then go back and clean up `CreateStatement` and `CreateExecutor`.

**Pull request merged**